### DISCUSSION TITLE
Typescript integration for Meteor packages

### DISCUSSION THREAD

**Discussion by piotrpospiech** [meteor/meteor#12080] [Feature Requests]
Hi, as a part of the Vazco & Meteor Software cooperation, I've just started working on native TypeScript type definitions support for Atmosphere packages and I want to share with you some details.

Currently, packages installed with Atmosphere are missing type definitions. Even if a source code would have them, it won't be used in your IDE. For this reason, we use DefinitelyTyped to support Atmosphere packages with type definitions. My goal is to remove the necessity to use an external library and import all types, that will be provided with the source code, to the project. That will help to keep TypeScript types up-to-date and improve a developer's experience.

The roadmap looks as follows:
1. **Proof of concept**
   - Atmosphere packages are located in the `~/.meteor/packages`. I will create an external script that will copy `index.d.ts` from the package source code to the project
   - It needs to define what package version is used on the project and it should localize the directory with its source code
   - At the beginning, it will be working with one specific package - `universe:i18n`
   - Proper TypeScript configuration will be needed, so I will provide documentation of what changes will be required in the `tsconfig.json`
2. **Release candidate**
   - Script should be integrated with Meteor to do it automatically. I still need to do some research about it, but there are the main ideas for now:
      - It can be done after starting the project. However, it probably has more disadvantages than advantages because it's time-consuming and not always needed.
      - Also it can be run after updating Atmosphere packages i.e., installing, updating, and removing.
   - Currently, TypeScript type definitions are stored in the DefinitelyTyped repository. A requirement for this to work natively in Meteor is that Atmosphere packages should have `index.d.ts` files in their main repository
3. **Stable release**
   - Script should find more than one file with type definitions. I will provide support for all `.d.ts` files in the package repository
   - More packages will be integrated with this feature

Feel free to comment and share your opinion.

References:
- [Atmosphere packages with typescript definitions?](https://forums.meteor.com/t/atmosphere-packages-with-typescript-definitions/54292/3)
- https://github.com/meteor/meteor/issues/10828
- https://github.com/meteor/meteor/issues/6177
- https://github.com/meteor/meteor/discussions/11735

**Comment by copleykj** [comment]
So I was actually giving this a bit of thought the other day myself and had very similar thoughts. Here was what I was thinking.

1. Meteor's build tool should package all .d.ts files and include them in the iso build bundle under a types directory
1. Create a CLI interface that can be called by things like npm lifecycle scripts such as `postinstall`
1. CLI should analyze `./meteor/versions` finding the currently used packages and versions
1. CLI should use Meteor's package cache, which now thanks to the build took contains a new types directory for each package version, and copy each package's types at the currently installed version (e.g. ~/.meteor/packages/univers_i18n/1.36.2/types)
1. CLI should place each type in a corresponding folder in `./node_modules/meteor' (e.g node_modules/meteor/universe:i18n/)
1. CLI should be called after `meteor add`, `meteor remove`, and `meteor update` (could we create some hooks in the meteor tool for this?) as well as by the 'postinstall` npm script.

There are some some caveats here like the fact that npm sometimes overwrites the node_modules dir in its entirety. This is mitigated by using the CLI to re-copy the types using the postinstall script, and other projects such as Prisma are using this approach to have a dynamically generated api that lives in the node_modules dir.

Hopefully this might give ideas that you might not have had yet. I'm very excited for this as this opens a whole realm of possibility for Meteor.

One other thing we might do is document a best practice for setting up the a package to generate the type defs before publishing. Maybe this could even be a "beforePublish" hook for the meteor tool that specifies a command to run before the publishing

```diff
  "meteor": {
    "mainModule": {
      "client": "client/main.tsx",
      "server": "server/main.ts"
    },
    "testModule": "tests/main.ts",
+   "hooks": {
+    "beforePublish": "tsc index.js --declaration --emitDeclarationOnly",
+    "afterPackageModification": "meteor-ts",
+   }
  },
```

**Reply by radekmie** [reply]
@copleykj Together with @piotrpospiech we've decided not to touch isopacks for now, as it'd require potentially backward-incompatible changes. That's why the plan is to rely solely on the package sources in `~/.meteor`.

**Reply by Julusian** [reply]
Instead of copying into `node_modules` and having to deal with issues when npm/yarn/pnpm cleans cleans up your additions, I think it would be less bug-prone to have it generate to a folder such as `/meteor/local/types` which can be referenced in the tsconfig https://www.typescriptlang.org/tsconfig#typeRoots property.

This will probably require users to ensure their tsconfig is setup appropriately, but that should be simple to document.

**Reply by copleykj** [reply]
All of this could very well become moot.. Maybe look to collaborate with @zodern on his recently released package? https://github.com/zodern/meteor-types

**Reply by fredmaiaarantes** [reply]
That would be ideal @copleykj. How can we join forces @zodern, @piotrpospiech, @radekmie?

**Reply by copleykj** [reply]
If possible I think it would be ideal if this were just part of the typescript package.

**Reply by piotrpospiech** [reply]
I reached out to @zodern to figure out how we can collaborate on this feature.

**Reply by fredmaiaarantes** [reply]
> I reached out to @zodern to figure out how we can collaborate on this feature. 

Nice!  Pls keep us posted.  Ideally, we should bring this feature into our TypeScript package and not require extra steps for developers.

**Reply by zodern** [reply]
I agree the integration should eventually be part of Meteor. I thought it would be good to initially experiment as a non-core package so we can try more solutions and better understand any problems first. Also, I have a few packages I want to add types to that support old Meteor versions that don't have the typescript package, so they would likely need a non-core solution anyway for publishing them.

Regarding [`zodern:types`](https://github.com/zodern/meteor-types), it seems to mostly work. The known issues are:
- The typescript language server used by vscode and some other code editors doesn't always detect when the types for a package are updated. Restarting the language server or code editor fixes it.
- Linter build plugins are slower than necessary and have some bugs. There are work arounds the package can use, and I also plan to submit a PR to Meteor to fix them
- For some local packages it uses a `.ts` instead of `.d.ts` file. It seems typescript has the package's `tsconfig.json` inherit parts of the app's `tsconfig.json`, which might make it more strict and create type errors in the package's code.

Things it is missing:
- types when importing a specific file from a package (for example: `import { addStory } from 'meteor/zodern:aurorae/react'`)
- setting up types for a package's dependencies on other Meteor packages so when running `tsc` for the package it can find the types
- I haven't looked into how typing the globals a package exports would work

I think the next steps are:
- Use `zodern:types` in apps and packages to see what issues come up
- Possibly explore other options of having typescript find a package's types. The code for the current implementation is in [writer.js](https://github.com/zodern/meteor-types/blob/master/writer.js)
- Decide how to pick which file to use for a package's types. The code for this is in [types-entry.js](https://github.com/zodern/meteor-types/blob/70a0aedba1a47d89c8f991d98ff72780eed90de4/types-entry.js#L7).
   - Decide where types should be within a package. When packages have `zodern:types` generate types, it currently does [this](https://github.com/zodern/meteor-types/blob/70a0aedba1a47d89c8f991d98ff72780eed90de4/publish-packages.js#L108).
- Figure out how to provide types for when an app imports a specific file from a package
- See what happens when Meteor packages use types from npm packages, and figure out how to better handle this and fix any issues with it.
- If we decide to stick with having a single `packages.d.ts` file, look into creating a source map for it so code editor's `Go to definition` feature jumps directly to the package's types instead of to the packages.d.ts file.

**Reply by filipenevola** [reply]
@zodern this is great.

I believe it would be better to have it as part of the core asap so core packages could also switch to this way of declaring types.

I don't see any issues, except for really old packages without typescript support, of having it in the core since day 0.

---

About VSCode, maybe you should talk with @matheusccastroo to see if [Meteor Toolbox Extension](https://marketplace.visualstudio.com/items?itemName=meteor-toolbox.meteor-toolbox) could help somehow.

**Reply by zodern** [reply]
We actually can start adding types to core packages now. You can do it by adding a single `.d.ts` file to the package, and add `typescript` as a dependency so Meteor includes the `.d.ts` file when publishing. A couple things we need to figure out for core packages to have types are:

1) How to have typescript prefer the types from core packages over those from @meteor/types. I had this working initially in `zodern:types`, but the approach it used caused other problems so I removed it.
2) How to type globals the package exports

We likely will run into other problems, so this would be a good way to check if the solution `zodern:types` uses would work, or if we need to find a different one. There is still a lot I don't know about typescript, but I am wondering if the current solution could only cover 80% - 90% of what we need. For example, I'm not sure it will be possible to add types for `module.hot`.

**Reply by radekmie** [reply]
I think I can help here:

> How to have typescript prefer the types from core packages over those from @meteor/types. I had this working initially in `zodern:types`, but the approach it used caused other problems so I removed it.

I simply wouldn't. We can move all `@types/meteor` to their corresponding packages and tell people to uninstall `@types/meteor`. Then, as a next step, integrate them in the code instead of having a single `.d.ts` file next to the source.

> How to type globals the package exports

That's solved in `@types/meteor` too. There's a [script](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/2de44476f971e3a1d36e6eb3a6c7a83708a7c79f/types/meteor/scripts/generate-globals), that takes all of the "standard modules" (i.e., the ones using [`declare module 'meteor/X'`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/2de44476f971e3a1d36e6eb3a6c7a83708a7c79f/types/meteor/meteor.d.ts#L5)) and transforms them into "globals" (i.e., uses [`declare namespace`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/2de44476f971e3a1d36e6eb3a6c7a83708a7c79f/types/meteor/globals/meteor.d.ts#L2)). And yes, you have to have both in order to make it work. _(Technically, we could try to make the global ones use modules, but that's kind of tricky; I already tried it for `@types/meteor` twice.)_

> For example, I'm not sure it will be possible to add types for `module.hot`.

Why? `@types/webpack-env` [does that](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/2de44476f971e3a1d36e6eb3a6c7a83708a7c79f/types/webpack-env/index.d.ts#L73).

**Comment by radekmie** [comment]
Starting [v2.8.1](https://docs.meteor.com/changelog.html#v28120221114) Meteor now offers its own types thanks to [zodern:types](https://github.com/zodern/meteor-types), see [migration steps](https://docs.meteor.com/changelog.html#migrationsteps-5) for more information.

**Comment by JanMP** [comment]
TS can infer types from JS files (and also use JSDoc). It would be really great if we could get the Coffeescript build tools to make this possible for Coffeescript as well. There's a VS-Code extension (CoffeeSense) that does that for CS to CS and TS to CS but not the other way round and it also does not work with 'meteor/package' imports at all.
